### ISSUE TITLE
Research Sassafras capability for software availability display

### ISSUE THREAD

**Issue by kevinkidwell** [cul-it/uls#753]

**Comment by kevinkidwell** [comment]
In researching this and seeing what we are able to get from Sassfras, we are able to get the software that is available by using their data-scope url. I'm using the [Mann Research Space](https://licensing.citlabs.cornell.edu/frame-core/software?setscope=eyJpZCI6IjljZWE2ZjI0ZGQ4ODI3NjQ5ZDllYjM1ZjE1YzAyZWI5IiwidHlwZSI6Im1hcCIsIm5hbWUiOiJNYW5uIExpYnJhcnkgUmVzZWFyY2ggU3BhY2UiLCJkZXNjIjoiQ29ybmVsbCBMYWJzIC8gTWFubiBMaWJyYXJ5IFJlc2VhcmNoIFNwYWNlIiwic2V0aWQiOiJzdGQiLCJzZXQiOiJDb3JuZWxsIExhYnMiLCJjc3QiOiJkaXZuIiwicGlkIjoiMHgwMDAwRkZCOSIsInBuYW1lIjoiRGVsZWdhdGVkT2JqZWN0cy5MSUJSQVJZLlVuaXRPYmplY3RzLlB1YmxpYy5EUyBDb21wdXRlcnMuTWFubi5SZXNlYXJjaCJ9) as an example.

The way that we are able to pull this in is as this long software list:
![Screen Shot 2020-04-02 at 1 49 11 PM](https://user-images.githubusercontent.com/29164102/78281493-c126c680-74e8-11ea-983c-a5fdd2848cfe.png)

This display is nowhere near what we currently have on Mann (which I think is much more user-friendly to see software at a glance):
![Screen Shot 2020-04-02 at 1 49 53 PM](https://user-images.githubusercontent.com/29164102/78281556-da2f7780-74e8-11ea-913d-a74140c24cb3.png)

**Comment by kevinkidwell** [comment]
Looking at the API, we can get a list of all the software using https://licensing.citlabs.cornell.edu/api/v2/product/items.

The problem is, the software doesn’t have associated divisions in the API. If we were able to ask Sassafras if it's possible to add that those division fields to be retrieved via the API, I think we'd be able to get what we need. We could then do something like https://licensing.citlabs.cornell.edu/api/v2/product/items?query=(division=65467) and it would give us only software that’s in Mann B30A.

**Comment by manolobevia** [comment]
Who can give credentials to access the API?

**Comment by cappadona** [comment]
@manolobevia I'm working on getting an application account with Pete and CIT Labs. For now if you want your own personal acct I'll send you a message in Slack with details.

**Comment by manolobevia** [comment]
@cappadona from Pete: I'll put in a request for access to CIT for access.  They are currently trying to set something up for Nick as well.
If you can help me getting a personal account that would be great. Thanks.

**Comment by cappadona** [comment]
@manolobevia You should request a personal acct from Pete for now and then follow the [API docs for authentication](https://www.sassafras.com/hrl/7.6/kr_restapi.html#authent).

I will post an update here when/if we are able to get an account with API access that we can use for our apps.

**Comment by cappadona** [comment]
Still working with Pat on getting a K2 account setup for use with our application that has appropriate permissions. He has a remote session scheduled with Sassafras support for this afternoon.

At the same time I have an active and ongoing thread with Sassafras support to work through the undocumented capabilities of the API. We're making progress and here's a preliminary list of the software that can be returned by the API for Mann Library (B30A, B30B, Kiosk, Research, Stone Classroom):

```js
[
  "7-Zip version 18",
  "7-Zip version 19",
  "Adobe Acrobat DC Pro",
  "Adobe Acrobat Reader DC Win",
  "Adobe Creative Cloud 2016",
  "Adobe Creative Cloud 2018",
  "Adobe Creative Cloud 2019"
  "Adobe Creative Cloud 2019",
  "Alertus Desktop 5 Win",
  "ArcGIS 10 Win",
  "ArcGIS Pro 2 Win",
  "AutoCAD 2020 Win",
  "Autodesk ReCap Pro 2019",
  "Deep Freeze 8",
  "EndNote X9 Win",
  "Evernote 6 Win",
  "FileZilla 3 Win",
  "Firefox 60 Win",
  "Firefox 68 Win",
  "Firefox 69 Win",
  "Firefox 72 Win",
  "Freedom Scientific JAWS 2018",
  "Geosoft Oasis Montaj",
  "Google Chrome 79 Win",
  "Google Earth 7 Win",
  "Internet Explorer 11 Win",
  "KeyVerify",
  "Maxon CINEMA 4D Lite R19 Win",
  "Microsoft Office 2016 Professional Plus Win",
  "Microsoft Office 2019 Professional Plus Win",
  "Microsoft Visual Studio 2019 Community Edition",
  "MiKTeX Console 2 Win",
  "MySQL Notifier 1.1",
  "MySQL Workbench 8.0 CE",
  "Node.js 4 Win",
  "Node.js 6 Win",
  "Node.js 8 Win",
  "Notepad++",
  "Palisade @RISK Industrial 7.0 Win",
  "Palisade @RISK Pro 7.6 Win",
  "Palisade the DecisionTools Suite 7.0 Win",
  "Palisade the DecisionTools Suite 7.6 Win",
  "PaperCut MF Client 17 Win",
  "QGIS Win",
  "Qt Designer 5 Win",
  "R Statistical Computing 3 Win",
  "Rhinoceros 6 Win",
  "RStudio Win",
  "SAS Base SAS 9 Win",
  "SAS Enterprise Guide 7 Win",
  "SAS JMP 13 Win",
  "SAS JMP Pro 14 Win",
  "SAS Studio 1 Win",
  "SAS/IML Studio 14 Win",
  "Stata/IC 15 Win",
  "Stata/IC 16 Win",
  "Tableau Desktop 2020 Win",
  "Tableau Public 2019",
  "VideoLAN VLC Player 2 Win",
  "VideoLAN VLC Player 3 Win",
]
``` 

@cul-it/clo-content-group Any initial thoughts when comparing to the existing [Mann software list](https://mannlib.cornell.edu/use/technology/software)? The Sassafras API does return metadata indicating which of these software titles apply to each of the 5 locations or machine types, but not much beyond that:

* B30A
* B30B
* Kiosk
* Research
* Stone Classroom

Also I'm not seeing the iMacs or Circ laptops in Sassafras at the moment, so I'll need to follow up with Desktop Services.

**Comment by cappadona** [comment]
Quick update: Confirmed with Pete on **2020-05-08** that the Sassafras client is not yet installed on Mann iMacs and Circ laptops.

**Comment by arh269** [comment]
Hi Nick,

Am I understanding you correctly that once we install the Sassafras client on Mann's iMacs and Circ laptops, we will be able to generate a list that shows us what software is available in a given location  such as Mann Library and which devices in that location it's available on, and that the issue is therefore whether or not a long list is adequate to our needs? If so, although I agree with Kevin that the grid we have now is more user-friendly, I think the list would work.

**Comment by sew268** [comment]
@cappadona @cul-it/clo-content-group I would agree with Andy that the list view for each location is sufficient.  Is there a possibility of removing software that is returned from the list so we don't show the public that KeyVerify for example, is on a particular group of machines?

**Comment by cappadona** [comment]
Thanks @arh269 & @sew268. There are several issues/questions at play here and I'll attempt to summarize from my perspective:

1. **Requirements/Specs**
   * My understanding is that our target is to reproduce existing functionality for the following units:
      * [Olin/Uris](https://olinuris.library.cornell.edu/spaces/computers/software) – managed via [CIT's Public Computing Mapping project](http://mapping.cit.cornell.edu/publiclabs/map/)
         * [Olin](http://mapping.cit.cornell.edu/publiclabs/map/view_data_point.cfm?data_point_id=752)
         * [Uris](http://mapping.cit.cornell.edu/publiclabs/map/view_data_point.cfm?data_point_id=755)
      * [Mann](https://mannlib.cornell.edu/use/technology/software) – managed via [CSV in GitHub repo](https://github.com/cul-it/mann-softwarelist-csv/)
   * Have we considered continuing with both of these existing solutions?
      * Is there a timeline for sunsetting CIT's mapping site and has Sassafras already been identified as the replacement?
      * The Mann software list was built as a [separate Vue app](https://github.com/cul-it/mann-wagon/tree/dev/vue/software-list) which we could easily embed into the new Wordpress site

1. **Sassafras native webapp views**
    * As mentioned previously, I've been working with Pat from CIT and Sassafras support to better understand Sassafras features as well as report several bugs that I have encountered. The good news is that many of these have been addressed with a minor `7.6.0.9` release that was distributed earlier this week and Pat upgraded our instance yesterday.
    * At the moment, we cannot list all of the software for multiple locations under a single unit, but I'm working on that with the Sassafras team
    * For each of Olin and Uris, we could link to a list of software for each location within the unit. Here's an example for Olin:
       * [Olin 1st Floor CoLab](https://licensing.citlabs.cornell.edu/software?setscope=eyJpZCI6IjJlM2NjNmI2M2IxNTY4NmU2ZjJmNWVmMDJjZjVkNjg0IiwidHlwZSI6Im1hcCIsIm5hbWUiOiJPbGluIDFzdCBGbG9vciBDb0xhYiIsImRlc2MiOiJDb3JuZWxsIExhYnMgLyBPbGluIDFzdCBGbG9vciBDb0xhYiIsInNldGlkIjoic3RkIiwic2V0IjoiQ29ybmVsbCBMYWJzIiwiY3N0IjoiZGl2biIsInBpZCI6IjB4MDAwMEZGNTAiLCJwbmFtZSI6IkRlbGVnYXRlZE9iamVjdHMuTElCUkFSWS5Vbml0T2JqZWN0cy5QdWJsaWMuRFMgQ29tcHV0ZXJzLk9saW4uMXN0IEZsb29yIENvTGFiIn0)
       * [Olin Library 1st Floor Productivity](https://licensing.citlabs.cornell.edu/software?setscope=eyJpZCI6IjBmN2Q1MmZkZjI2YTdkNWU0NTU5ZjRkNTMwZTMwZmM2IiwidHlwZSI6Im1hcCIsIm5hbWUiOiJPbGluIExpYnJhcnkgMXN0IEZsb29yIFByb2R1Y3Rpdml0eSIsImRlc2MiOiJDb3JuZWxsIExhYnMgLyBPbGluIExpYnJhcnkgMXN0IEZsb29yIFByb2R1Y3Rpdml0eSIsInNldGlkIjoic3RkIiwic2V0IjoiQ29ybmVsbCBMYWJzIiwiY3N0IjoiZGl2biIsInBpZCI6IjB4MDAwMEZGNkMiLCJwbmFtZSI6IkRlbGVnYXRlZE9iamVjdHMuTElCUkFSWS5Vbml0T2JqZWN0cy5QdWJsaWMuRFMgQ29tcHV0ZXJzLk9saW4uMXN0IEZsb29yIFByb2R1Y3Rpdml0eSJ9)
       * [Olin Library 1st Floor Scanning Stations](https://licensing.citlabs.cornell.edu/software?setscope=eyJpZCI6IjY5ODM4YWVmMzUxYWUwNTc4OWMyNzRmYjlkMjU1OGE5IiwidHlwZSI6Im1hcCIsIm5hbWUiOiJPbGluIExpYnJhcnkgMXN0IEZsb29yIFNjYW5uaW5nIFN0YXRpb25zIiwiZGVzYyI6IkNvcm5lbGwgTGFicyAvIE9saW4gTGlicmFyeSAxc3QgRmxvb3IgU2Nhbm5pbmcgU3RhdGlvbnMiLCJzZXRpZCI6InN0ZCIsInNldCI6IkNvcm5lbGwgTGFicyIsImNzdCI6ImRpdm4iLCJwaWQiOiIweDAwMDBGRjY0IiwicG5hbWUiOiJEZWxlZ2F0ZWRPYmplY3RzLkxJQlJBUlkuVW5pdE9iamVjdHMuUHVibGljLkRTIENvbXB1dGVycy5PbGluLjFzdCBGbG9vciBTY2FubmluZyBTdGF0aW9ucyJ9)
       * [Olin Library 5th floor Grad lounge](https://licensing.citlabs.cornell.edu/software?setscope=eyJpZCI6IjcyNzdmZmFmNGU4OTBhZjE0NTkxYmRlZGJiMTM5ZTFjIiwidHlwZSI6Im1hcCIsIm5hbWUiOiJPbGluIExpYnJhcnkgNXRoIGZsb29yIEdyYWQgTG91bmdlIiwiZGVzYyI6IkNvcm5lbGwgTGFicyAvIE9saW4gTGlicmFyeSA1dGggZmxvb3IgR3JhZCBMb3VuZ2UiLCJzZXRpZCI6InN0ZCIsInNldCI6IkNvcm5lbGwgTGFicyIsImNzdCI6ImRpdm4iLCJwaWQiOiIweDAwMDBGRjZCIiwicG5hbWUiOiJEZWxlZ2F0ZWRPYmplY3RzLkxJQlJBUlkuVW5pdE9iamVjdHMuUHVibGljLkRTIENvbXB1dGVycy5PbGluLjV0aCBmbG9vciBQcm9kdWN0aXZpdHkifQ)
       * [Olin Library Basement - B48](https://licensing.citlabs.cornell.edu/software?setscope=eyJpZCI6IjZkNjc2ODRkYjA5MGI2YjJlM2RjNDU1YjEwZTFlYWY5IiwidHlwZSI6Im1hcCIsIm5hbWUiOiJPbGluIExpYnJhcnkgQmFzZW1lbnQgLSBCNDgiLCJkZXNjIjoiQ29ybmVsbCBMYWJzIC8gT2xpbiBMaWJyYXJ5IEJhc2VtZW50IC0gQjQ4Iiwic2V0aWQiOiJzdGQiLCJzZXQiOiJDb3JuZWxsIExhYnMiLCJjc3QiOiJkaXZuIiwicGlkIjoiMHgwMDAwRkY2MyIsInBuYW1lIjoiRGVsZWdhdGVkT2JqZWN0cy5MSUJSQVJZLlVuaXRPYmplY3RzLlB1YmxpYy5EUyBDb21wdXRlcnMuT2xpbi5CYXNlbWVudCJ9)
       * [Olin Library Stacks - Stand up Kiosks](https://licensing.citlabs.cornell.edu/software?setscope=eyJpZCI6IjM0ZDgxODUxNWZkOTk4ZWQ5YzFjMjUxMzMzNzUzY2VhIiwidHlwZSI6Im1hcCIsIm5hbWUiOiJPbGluIExpYnJhcnkgU3RhY2tzIC0gU3RhbmQgdXAgS2lvc2tzIiwiZGVzYyI6IkNvcm5lbGwgTGFicyAvIE9saW4gTGlicmFyeSBTdGFja3MgLSBTdGFuZCB1cCBLaW9za3MiLCJzZXRpZCI6InN0ZCIsInNldCI6IkNvcm5lbGwgTGFicyIsImNzdCI6ImRpdm4iLCJwaWQiOiIweDAwMDBGRjYyIiwicG5hbWUiOiJEZWxlZ2F0ZWRPYmplY3RzLkxJQlJBUlkuVW5pdE9iamVjdHMuUHVibGljLkRTIENvbXB1dGVycy5PbGluLlN0YWNrcyJ9)
   * Does this suffice?
   * Another potential option would be to link the Sassafras map with Olin highlighted as shown below, but I'm working with the support team to see if that's possible
      ![Screenshot_2020-05-15 Maps -- Computer Availability -- Cornell Labs](https://user-images.githubusercontent.com/409842/82066421-61057180-969d-11ea-9535-f277d42e172b.png)

1. **Sassafras API**
   * If we are indeed attempting to replicate existing functionality at Mann, and the decision has been made to not continue with the CSV and Vue app, the Sassafras API has potential to provide the data needed to power a refactored Vue app
   * There are a couple caveats at the moment that I'm still working through with Sassafras support:
      * We cannot hide/restrict any particular piece of software from being returned (answering @sew268's question [from above](https://github.com/cul-it/uls/issues/753#issuecomment-627560131))
      * We cannot provide alternate names for the software (see list of unique software titles at Mann [from above](https://github.com/cul-it/uls/issues/753#issuecomment-625249666))
   * If either of these are dealbreakers and it turns out that Sassafras cannot provide these features prior to launch, an alternative is to manage each of these on our end by curating the data feed from the API within the code (suppressing certain software titles and editing others)
      * Please don't confuse this suggestion as an endorsement or recommendation on my part 😉

**Comment by cappadona** [comment]
Thanks for the discussion today @cul-it/clo-content-group.

Here are the action items I captured:

- Use existing Vue app and CSV for initial Mann implementation (#820)
- Follow up with Olin/Uris stakeholders if they are good with following that path as well (#817)
- Follow up with desktop services group to determine life expectancy of Sassafras (#818)

I also attempted to distill the requirements for software availability based on our discussion, which you can find in #816. I encourage everyone to revise as needed.

My main takeaway in regards to the task at hand in assessing Sassafras is that we should temporarily place this on hold as we wait on how some of the above action items play out.

**Comment by ajs17** [comment]
According to Pete Magnus via private slack message on 5/19/2020: "My understanding is that there are labs on campus they will still be supporting.  The units will be charged a fee for the lab support.  My understanding is that they will continue to need the Sassafras service for the Apps on Demand licensing."

**Comment by cappadona** [comment]
While the Sassafras K2 API doesn't support all of our needs natively, I was able to work with Pat (locally) and Sassafras engineers, who supplied a custom format file that will render our reports as JSON when requesting reports from the webapp using the following URL pattern:

```
/archive/{{reportKSR}}?untagged=lib-hide&format=transform-cornell-products
```

Technically, we're not retrieving this data via the API, but the end result is that we can manage Library software in Key Reporter without needing to model any metadata locally in our websites.

More specifically, this approach allows us to:

1. Suppress software from our website by tagging products (aka software) in the Key Reporter web application
1. Override the name of any particular product using the Notes field in the Key Reporter web application

Here's a snippet of the array (first two items) that is returned:

```json
[
  {
    "id": "5A55ADEF8ECB41A398616A659BD463A8",
    "title": "7-Zip version 19",
    "division": "DelegatedObjects.LIBRARY.UnitObjects.Public.DS Computers.Mann.B30A",
    "familyname": "7-Zip"
  },
  {
    "id": "5A55ADEF9F704036B8810CF5DEAEA704",
    "title": "Adobe Acrobat Reader DC Win",
    "division": "DelegatedObjects.LIBRARY.UnitObjects.Public.DS Computers.Mann.B30A",
    "familyname": "Adobe Acrobat Reader"
  }
  ...
]
```

The `title` field dynamically returns the value from the `Notes` field if it is populated, otherwise it defaults to the `Name` field.

The `familyname` field is another alternative we could use for labeling software on our sites, but the Sassafras engineers warned me that this product family feature is not fully baked and we're likely to encounter products that are not assigned to a family.

---

For the time being, I've set up three report templates which can be run on demand from widgets I added to the dashboard for `ReportAPIuser`.
* Mann Library Software
* Olin Library Software
* Uris Library Software

> we can change this up to run on a set schedule if we so choose

**Comment by cappadona** [comment]
## February 2021 Update

I needed to reach out to Pat Washburn yesterday to get the custom format file provided by Sassafras back on our server. Apparently it was lost/deleted during a recent server update. Our software availability app is not functional without this file in place.
